lib/widgets/action_sheet.dart

```diff
@@ -22,6 +23,297 @@ import 'store.dart';
+        buttons.add(unmute);
+      case UserTopicVisibilityPolicy.none:
+      case UserTopicVisibilityPolicy.unmuted:
+        buttons.add(mute);
+      case UserTopicVisibilityPolicy.followed:
+        buttons.add(mute);
+      case UserTopicVisibilityPolicy.unknown:
+    }
+  } else if (channelMuted != null && channelMuted) {
+    if (supportsUnmutingTopics) {
+      switch (visibilityPolicy) {
+        case UserTopicVisibilityPolicy.none:
+        case UserTopicVisibilityPolicy.muted:
+          buttons.add(unmuteInMutedChannel);
+        case UserTopicVisibilityPolicy.unmuted:
+          buttons.add(mute);
+        case UserTopicVisibilityPolicy.followed:
+          buttons.add(mute);
+        case UserTopicVisibilityPolicy.unknown:
+      }
+    }
+  }
+
+  if (supportsFollowingTopics) {
+    switch (visibilityPolicy) {
+      case UserTopicVisibilityPolicy.muted:
+      case UserTopicVisibilityPolicy.none:
+      case UserTopicVisibilityPolicy.unmuted:
+        buttons.add(follow);
+      case UserTopicVisibilityPolicy.followed:
+        buttons.add(unfollow);
+      case UserTopicVisibilityPolicy.unknown:
+    }
+  }
+
+  // TODO(server-7): remove
+  if (buttons.isEmpty) {
+    // Modern Zulip servers should always support at least some topic actions.
+    assert(store.connection.zulipFeatureLevel! < 170);
+    return;
+  }
+
+  showModalBottomSheet<void>(
+    context: context,
+    // Clip.hardEdge looks bad; Clip.antiAliasWithSaveLayer looks pixel-perfect
+    // on my iPhone 13 Pro but is marked as "much slower":
+    //   https://api.flutter.dev/flutter/dart-ui/Clip.html
+    clipBehavior: Clip.antiAlias,
+    useSafeArea: true,
+    isScrollControlled: true,
+    builder: (BuildContext _) {
+      return SafeArea(
+        minimum: const EdgeInsets.only(bottom: 16),
+        child: Padding(
+          padding: const EdgeInsets.fromLTRB(16, 0, 16, 0),
+          child: Column(
+            crossAxisAlignment: CrossAxisAlignment.stretch,
+            mainAxisSize: MainAxisSize.min,
+            children: [
+              // TODO(#217): show message text
+              Flexible(child: InsetShadowBox(
+                top: 8, bottom: 8,
+                color: DesignVariables.of(context).bgContextMenu,
+                child: SingleChildScrollView(
+                  padding: const EdgeInsets.only(top: 16, bottom: 8),
+                  child: ClipRRect(
+                    borderRadius: BorderRadius.circular(7),
+                    child: Column(spacing: 1,
+                      children: buttons))))),
+              const ActionSheetCancelButton(),
+            ])));
+    });
+}
+
+abstract class TopicActionSheetButton extends ActionSheetMenuItemButton {
+  const TopicActionSheetButton({
+    super.key,
+    required this.streamId,
+    required this.topic,
+    required this.topicParentContext,
+  });
+
+  final int streamId;
+  final String topic;
+  final BuildContext topicParentContext;
```

Comment:
This name doesn't feel right conceptually. I'll see about sending a quick PR refactoring the existing message-action-sheet options that might help clarify what this field is about.

Reply:
Sent #1044.